### PR TITLE
Update RoT bootloader, a.k.a. stage0

### LINKED ISSUE #1043: Update RoT memory map to reserve RAM for updates
Due to system constraints there is no alternate image for the stage0 bootloader. This makes updating it a riskier proposition than Hubris since a broken stage0 may be equivalent to a bricked board. To reduce risk, we can reserve a region of RAM for staging an update before committing to flash.

Section 2.1.8 of the LPC55 manual gives a table of RAM layouts. The final product will use the 320 KB layout. We have a size limit of 64K for stage0. We've been testing with the 256KB layout which lacks RAM3 and we aren't running into any limitations so reserving RAM3 looks plausible.

### PR DESCRIPTION
**Pull request by lzrd**
This PR has four commits to simplify reviewing.

  - The first is Laura's macro to manipulate the HASHCRYPT interrupt vector.
  - 2nd is pre-main kernel changes to create a staging area for Bootleby update and to validate the (now) four RoT flash banks and record those results for use by update_server.
  - 3rd are the API changes to support stage0 update including treating the bootloader as a separate component with A (active) and B (stage0next) banks.
  - last is the code in the RoT update_server to actually update stage0

Before merging this PR, Cargo.toml needs to be edited to reference the gateway-message-service main branch.
Merging needs to be coordinated with an MGS merge.

Closes #1043, #1404, #1548, #1353,

### CODE REVIEW

**Comment by lzrd** [comment]
See https://github.com/oxidecomputer/sprot-e2e for test scripts

File: Cargo.toml [hunk 1]
```diff
@@ -119,7 +119,9 @@ zip = { version = "0.6", default-features = false, features = ["bzip2"] }
 # Oxide forks and repos
 attest-data = { git = "https://github.com/oxidecomputer/dice-util", default-features = false, version = "0.1.0" }
 dice-mfg-msgs = { git = "https://github.com/oxidecomputer/dice-util", default-features = false, version = "0.2.1" }
-gateway-messages = { git = "https://github.com/oxidecomputer/management-gateway-service", default-features = false, features = ["smoltcp"] }
+#gateway-messages = { git = "https://github.com/oxidecomputer/management-gateway-service", default-features = false, features = ["smoltcp"] }
+# XXX
+gateway-messages = { git = "https://github.com/oxidecomputer/management-gateway-service", branch="update-stage0", default-features = false, features = ["smoltcp"] }
```
Comment (lzrd) [inline Cargo.toml:124]: This must be fixed before merge.

File: drv/lpc55-update-server/src/main.rs [hunk 2]
```diff
@@ -596,6 +624,133 @@ impl ServerImpl<'_> {
...
+        // signature routine.
+
+        // Read stage0next contents into RAM.
+        let staged = image_range(RotComponent::Stage0, SlotId::B);
+        let len = self.read_flash_image_to_cache(staged.0)?;
+        let bootloader = &self.fw_cache[..len / core::mem::size_of::<u32>()];
+
+        let mut hash = Sha3_256::new();
+        for page_start in (0..(bootloader.len() * 4)).step_by(512) {
+            // XXX
```
Comment (hawkw) [inline drv/lpc55-update-server/src/main.rs:691]: hmm?

File: drv/lpc55-update-server/src/main.rs [hunk 3]
```diff
@@ -573,6 +623,411 @@ impl ServerImpl<'_> {
...
+            hash.update(
+                bootloader.as_bytes()[page_start..page_start + 512].as_ref(),
+            );
+        }
```
Comment (hawkw) [inline drv/lpc55-update-server/src/main.rs:695]: style nit: i think we could loop over every 512-byte page by writing this, instead:
```suggestion
        for page in bootloader.as_bytes().chunks(512) {
            hash.update(page);
        }
```

which feels much clearer to me?

File: drv/lpc55-update-server/src/main.rs [hunk 4]
```diff
@@ -615,6 +770,137 @@ impl ServerImpl<'_> {
...
+
+    fn read_flash_image_to_cache(
+        &mut self,
+        span: Range<u32>,
+    ) -> Result<usize, UpdateError> {
+        // Returns error if flash page is erased.
+        let staged = image_range(RotComponent::Stage0, SlotId::B);
+        let len = self.valid_flash_image_present(&staged.0)?;
+        if len as u32 > span.end || len > self.fw_cache.as_bytes().len() {
+            return Err(UpdateError::BadLength); // XXX unique error
```
Comment (hawkw) [inline drv/lpc55-update-server/src/main.rs:861]: do we want to sort out these error types before merging this PR?

File: drv/lpc55-update-server/src/main.rs [hunk 5]
```diff
@@ -573,6 +623,411 @@ impl ServerImpl<'_> {
...
+                return Err(UpdateError::ImageMismatch);
+            }
+        }
+        Ok(())
+    }
+
+    // Looking at a region of flash, determine if there is a possible NXP
+    // image programmed. Return the length in bytes of the flash pages
+    // comprising the image including padding to fill to a page boundary.
+    fn valid_flash_image_present(
```
Comment (hawkw) [inline drv/lpc55-update-server/src/main.rs:813]: nit, take it or leave it: it feels a bit strange to me that a method name ending in `_present` returns a number and not a boolean? `_present` feels like a bool to me. but, i'm open to being convinced otherwise.

File: drv/lpc55-update-server/src/main.rs [hunk 6]
```diff
@@ -573,6 +623,411 @@ impl ServerImpl<'_> {
...
+                    // but this is enough bytes for an NXP header and not
+                    // bigger than the flash slot.
+                    if len as usize <= span.len() && len >= HEADER_OFFSET {
+                        return Ok(len as usize);
+                    }
+                }
+                Err(UpdateError::BadLength)
+            }
+            Err(error) => Err(error),
+        }
```
Comment (hawkw) [inline drv/lpc55-update-server/src/main.rs:835]: style nit: can't this `match` be:
```suggestion
         indirect_flash_read(
            &self.flash,
            span.start + LENGTH_OFFSET as u32,
            buf[..].as_bytes_mut(),
        )?;
 
        if let Some(len) = round_up_to_flash_page(buf[0]) {
            // The minimum image size should be further constrained
            // but this is enough bytes for an NXP header and not
            // bigger than the flash slot.
            if len as usize <= span.len() && len >= HEADER_OFFSET {
                return Ok(len as usize);
            }
        }
        
        Err(UpdateError::BadLength)
```

File: drv/lpc55-update-server/src/main.rs [hunk 7]
```diff
@@ -573,6 +623,411 @@ impl ServerImpl<'_> {
...
+    }
+
+    fn valid_cache_image_present(&self) -> Result<usize, UpdateError> {
+        let len = if let Some(len) = round_up_to_flash_page(
+            self.fw_cache[LENGTH_OFFSET / core::mem::size_of::<u32>()],
+        ) {
+            len
+        } else {
+            return Err(UpdateError::BadLength);
+        };
```
Comment (hawkw) [inline drv/lpc55-update-server/src/main.rs:845]: style nit: the if-else feels a bit hard to follow due to the weird way `rustfmt` wants to format this,what do you think of:
```suggestion
        let len =  round_up_to_flash_page(
            self.fw_cache[LENGTH_OFFSET / core::mem::size_of::<u32>()],
        ).ok_or(UpdateError::BadLength)?;
```

**Review by hawkw — COMMENTED** [review]
I'm not very familiar with the details of the LPC55 platform, so my review focuses more or less exclusively on the implementation. I've mostly left some code style suggestions, which you're welcome to ignore --- I hope they're helpful.

File: drv/lpc55-update-server/src/main.rs [hunk 8]
```diff
@@ -273,14 +294,55 @@ impl idl::InOrderUpdateImpl for ServerImpl<'_> {
...
+            0 => Err(UpdateError::VersionNotSupported.into()),
+            1 => Ok(VersionedRotBootInfo::V1(RotBootInfo {
+                active: boot_state.active.into(),
+                persistent_boot_preference,
+                pending_persistent_boot_preference,
+                transient_boot_preference,
+                slot_a_sha3_256_digest: Some(boot_state.a.digest),
+                slot_b_sha3_256_digest: Some(boot_state.b.digest),
+            })),
+            _ => Ok(VersionedRotBootInfo::V2(RotBootInfoV2 {
```
[the remark this answers is not in the document]
Comment (lzrd) [reply drv/lpc55-update-server/src/main.rs:338]: This is forward compatibility. If there is a next version that this implementation doesn't know about it will return the highest that it does know about. That allows the control plane to continue the update procedure without another roundtrip.

[the remark this answers is not in the document]
Comment (hawkw) [reply drv/lpc55-update-server/src/main.rs:338]: Ah, I see --- that makes sense. Maybe a comment explaining that would be nice?

File: app/gimletlet/app-mgmt.toml [hunk 9]
```diff
@@ -34,7 +34,7 @@ task-slots = ["jefe"]
 name = "drv-user-leds"
 features = ["stm32h7"]
 priority = 2
-max-sizes = {flash = 2048, ram = 1024}
+max-sizes = {flash = 2048, ram = 2048}
```
[the remark this answers is not in the document]
Comment (lzrd) [reply app/gimletlet/app-mgmt.toml:37]: This app builds fine for me.
I've adjusted or remove a few to not break things that compiled before. And I think removing the unnecessary `max-sizes` and `required` settings in the app.toml files is a good idea. Let's do that in a global cleanup PR though.

File: lib/stage0-handoff/src/rot_update_details.rs [hunk 10]
```diff
@@ -30,22 +67,112 @@ pub struct RotBootState {
...
+    }
+}
+
+fits_in_ram!(RotBootStateV2);
+
+#[derive(
+    Debug, Copy, Clone, PartialEq, Eq, Deserialize, Serialize, SerializedSize,
 )]
 pub struct RotImageDetails {
+    // The SHA3-256 measurement of all programmed pages in the flash slot.
```
[the remark this answers is not in the document]
Comment (lzrd) [reply lib/stage0-handoff/src/rot_update_details.rs:130]: done

File: lib/stage0-handoff/src/rot_update_details.rs [hunk 11]
```diff
@@ -30,22 +67,112 @@ pub struct RotBootState {
...
+    /// 0xff bytes. All non-erased pages in the flash slot are included in the
+    /// digest.
+    Sha3_256([u8; 32]),
+}
+
+#[derive(
+    Debug, Copy, Clone, PartialEq, Eq, Deserialize, Serialize, SerializedSize,
+)]
+pub struct RotImageDetailsV2 {
+    pub digest: [u8; 32],
```
[the remark this answers is not in the document]
Comment (lzrd) [reply lib/stage0-handoff/src/rot_update_details.rs:171]: done

File: app/gimletlet/app-mgmt.toml (hunk 9 above)
[the remark this answers is not in the document]
Comment (hawkw) [reply app/gimletlet/app-mgmt.toml:37]: sounds good to me!

File: lib/lpc55-rot-startup/src/images.rs [hunk 12]
```diff
@@ -55,124 +55,341 @@ extern "C" {
...
+    // page, are a possibility that could be a problem with respect to
+    // unexpected crashes or catching exfiltration attempts.
+    fn new(
+        slot: &FlashSlot,
+        run: Range<u32>,
+        header_required: bool,
+        syscon: &SYSCON,
+    ) -> Result<Image, ImageError> {
+        // Make sure we can access the page where the vectors live.
+        // SAFETY: Link time constants from our own image.
```
[the remark this answers is not in the document]
Comment (lzrd) [reply lib/lpc55-rot-startup/src/images.rs:188]: I count 43  SAFETY and 79 Safety with some inconsistent use of a colon after that word.
Making this uniform throughout Hubris should be a separate cleanup commit.
I'll switch the ones here to "Safety: " to side with the majority bring the score to 28 SAFETY and 90 Safety.

[the remark this answers is not in the document]
Comment (hawkw) [reply lib/lpc55-rot-startup/src/images.rs:188]: sounds good to me --- definitely not expecting the global cleanup pass to happen in this PR!

File: lib/lpc55-rot-startup/src/images.rs [hunk 13]
```diff
@@ -55,124 +55,341 @@ extern "C" {
...
+            0 => clear_bit!(syscon.presetctrl0, PMASK),
+            1 => clear_bit!(syscon.presetctrl1, PMASK),
+            2 => clear_bit!(syscon.presetctrl2, PMASK),
+            _ => panic!(),
+        }
+
+        let authorized = unsafe {
+            lpc55_romapi::authenticate_image(self.span.start).is_ok()
+        };
+        // let authorized = true;
```
[the remark this answers is not in the document]
Comment (lzrd) [reply lib/lpc55-rot-startup/src/images.rs:379]: The comment? Gone.

File: drv/lpc55-update-server/src/main.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (lzrd) [reply drv/lpc55-update-server/src/main.rs:338]: done

File: drv/lpc55-update-server/src/main.rs (hunk 3 above)
Comment (lzrd) [reply drv/lpc55-update-server/src/main.rs:695]: Thanks.

File: drv/ksz8463/Cargo.toml [hunk 14]
```diff
@@ -6,6 +6,7 @@ edition = "2021"
 [dependencies]
 num-traits = { workspace = true }
 
+counters = {path = "../../lib/counters"}
```
Comment (mkeeter) [inline drv/ksz8463/Cargo.toml:9]: Is this change (and the one to `drv/ksz8463/src/lib.rs`) intentional?  It looks like a merge / rebase glitch.

File: drv/lpc55-update-api/src/lib.rs [hunk 15]
```diff
@@ -109,6 +191,24 @@ pub enum UpdateTarget {
     Bootloader,
 }
 
+/// Designate a logical sub-component of the RoT
+///
```
Comment (mkeeter) [inline drv/lpc55-update-api/src/lib.rs:195]:
```suggestion
```

File: drv/stm32h7-spi-server-core/Cargo.toml [hunk 16]
```diff
@@ -14,6 +14,7 @@ zerocopy = { workspace = true }
 drv-spi-api = { path = "../spi-api" }
 drv-stm32h7-spi = { path = "../stm32h7-spi" }
 drv-stm32xx-sys-api = { path = "../stm32xx-sys-api" }
+counters = { path = "../../lib/counters" }
```
Comment (mkeeter) [inline drv/stm32h7-spi-server-core/Cargo.toml:17]: Same here, these changes (and the one to `stm32h7-spi-server-core/src/lib.rs`) seem accidentally included in the PR

File: drv/vsc-err/Cargo.toml [hunk 17]
```diff
@@ -4,6 +4,7 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
+counters = { path = "../../lib/counters" }
```
Comment (mkeeter) [inline drv/vsc-err/Cargo.toml:7]: same here

File: drv/stm32h7-spi-server-core/Cargo.toml (hunk 16 above)
Comment (lzrd) [reply drv/stm32h7-spi-server-core/Cargo.toml]: Those were the result of me trying to rebase without having to do a push --force. The point there being that it would not invalidate previous comments from github. I've done a normal rebase and forced that. These should go away but I will give it a look.

File: lib/lpc55-hashcrypt/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+/// proc macro to generate iocon function for LPC55
+extern crate proc_macro;
+use proc_macro::TokenStream;
+extern crate proc_macro2;
+use quote::quote;
+
+#[proc_macro]
+pub fn dynamic_hashcrypt(_item: TokenStream) -> TokenStream {
```
[the remark this answers is not in the document]
Comment (labbott) [reply lib/lpc55-hashcrypt/src/lib.rs:12]: I wrote this code many months ago and I'm doing my best to remember why this needed to be a proc macro. I _think_ there was some issue with calling `kern::arch::DefaultHandler()`. I'll dig through my notes some more and see if I can remember why.

File: lib/lpc55-hashcrypt/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    let code = quote! {
+        static mut USE_ROM: core::sync::atomic::AtomicBool = core::sync::atomic::AtomicBool::new(false);
```
[the remark this answers is not in the document]
Comment (labbott) [reply lib/lpc55-hashcrypt/src/lib.rs:14]: My poor understanding of the Atomic API!

File: lib/lpc55-rot-startup/src/images.rs [hunk 20]
```diff
@@ -55,124 +55,340 @@ extern "C" {
...
+    fn check_signature(&self, syscon: &SYSCON) -> Result<(), ImageError> {
+        const HASHAES: u32 = 32 + 32 + 18;
+        const PMASK: u32 = 1 << (HASHAES % 32);
+        const REG_NUM: u32 = HASHAES / 32; // XXX must be 0, 1, or 2
+        match REG_NUM {
+            0 => clear_bit!(syscon.presetctrl0, PMASK),
+            1 => clear_bit!(syscon.presetctrl1, PMASK),
+            2 => clear_bit!(syscon.presetctrl2, PMASK),
+            _ => panic!(),
+        }
```
Comment (labbott) [inline lib/lpc55-rot-startup/src/images.rs:374]: We shouldn't need to use this manual bit entry, we can just use the pac APIs as intended https://docs.rs/lpc55-pac/latest/lpc55_pac/syscon/presetctrl0/index.html etc.

File: lib/lpc55-hashcrypt/src/lib.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (hawkw) [reply lib/lpc55-hashcrypt/src/lib.rs:14]: Well, in that case, this could probably be changed to 
```suggestion
        static USE_ROM: core::sync::atomic::AtomicBool = core::sync::atomic::AtomicBool::new(false);
```
and the use of `unsafe` can be removed!

File: lib/lpc55-hashcrypt/src/lib.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (labbott) [reply lib/lpc55-hashcrypt/src/lib.rs:12]: I think you are right that this code need not be procedural. The procedural bit was probably a holdover from some earlier attempts.

File: drv/lpc55-update-server/src/main.rs [hunk 21]
```diff
@@ -432,7 +403,7 @@ impl idl::InOrderUpdateImpl for ServerImpl<'_> {
             return Err(UpdateError::UpdateInProgress.into());
         }
         self.syscon.chip_reset();
-        panic!()
+        unreachable!();
```
Comment (labbott) [inline drv/lpc55-update-server/src/main.rs:406]: Did this reduce the flash size? according to https://doc.rust-lang.org/stable/std/macro.unreachable.html this will still produce an implicity panic

File: drv/lpc55-update-server/src/main.rs [hunk 22]
```diff
@@ -573,6 +628,406 @@ impl ServerImpl<'_> {
...
+        indirect_flash_read(
+            &self.flash,
+            span.start,
+            self.fw_cache[0..len / core::mem::size_of::<u32>()].as_bytes_mut(),
+        )?;
+        Ok(len)
+    }
+
+    // XXX This API needs to be sorted out.
+    // The lower level routines are being replaced in PR 1500
```
Comment (labbott) [inline drv/lpc55-update-server/src/main.rs:872]: Is this still true?

File: drv/lpc55-update-server/src/main.rs [hunk 23]
```diff
@@ -573,6 +628,406 @@ impl ServerImpl<'_> {
...
+    fn write_cache_to_flash(
+        &mut self,
+        component: RotComponent,
+        slot: SlotId,
+    ) -> Result<(), UpdateError> {
+        let clen = self.cache_image_len()?;
+        if clen % BYTES_PER_FLASH_PAGE != 0 {
+            return Err(UpdateError::BadLength); // XXX need unique error
```
Comment (labbott) [inline drv/lpc55-update-server/src/main.rs:880]: Still need the unique error?

File: drv/lpc55-update-server/src/main.rs [hunk 24]
```diff
@@ -963,6 +1343,18 @@ fn copy_from_flash_range(
...
+    // Safety: Data is published by stage0
+    let addr = unsafe { BOOTSTATE.assume_init_ref() };
+    RotBootStateV2::load_from_addr(addr)
+}
+
+fn round_up_to_flash_page(offset: u32) -> Option<u32> {
+    offset
+        .checked_add(BYTES_PER_FLASH_PAGE as u32 - 1)
+        .map(|sum| sum & !(BYTES_PER_FLASH_PAGE as u32 - 1))
+}
```
Comment (labbott) [inline drv/lpc55-update-server/src/main.rs:1356]: Can we use https://doc.rust-lang.org/stable/std/primitive.u32.html#method.checked_next_multiple_of ? This is a new stable API

File: lib/lpc55-hashcrypt/src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+/// proc macro to generate iocon function for LPC55
+extern crate proc_macro;
+use proc_macro::TokenStream;
+extern crate proc_macro2;
+use quote::quote;
+
+#[proc_macro]
```
Comment (labbott) [inline lib/lpc55-hashcrypt/src/lib.rs:11]: Can we delete this code if we're not using it?

File: lib/lpc55-rot-startup/src/lib.rs [hunk 26]
```diff
@@ -290,3 +323,5 @@ extern "C" fn nuke_stack() {
         )
     }
 }
+
+lpc55_hashcrypt::dynamic_hashcrypt!();
```
Comment (labbott) [inline lib/lpc55-rot-startup/src/lib.rs:327]: Do we we need this or can we go to the non-procmacro version

File: drv/lpc55-update-server/src/main.rs (hunk 23 above)
Comment (lzrd) [reply drv/lpc55-update-server/src/main.rs:880]: I decided not. If some length in the image is invalid its not hard to figure out which one.

File: lib/lpc55-hashcrypt/src/lib.rs (hunk 25 above)
Comment (lzrd) [reply lib/lpc55-hashcrypt/src/lib.rs:11]: This code has been previously removed.

File: lib/lpc55-rot-startup/src/lib.rs (hunk 26 above)
Comment (lzrd) [reply lib/lpc55-rot-startup/src/lib.rs:327]: Already removed

**Review by aapoalas — COMMENTED** [review]
Random questions, please excuse my curiosity.

**Comment by labbott** [comment]
LGTM! I'll give the final approval once the MGS portion is merged

File: drv/lpc55-update-server/src/images.rs [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+    static __this_image: [u32; 0];
+}
+
+// Location of the NXP header
+pub const HEADER_BLOCK: usize = 0;
+
+// NXP LPC55's mixed header/vector table offsets
+const RESET_VECTOR_OFFSET: usize = 0x04;
+pub const LENGTH_OFFSET: usize = 0x20;
+pub const HEADER_OFFSET: u32 = 0x130;
```
[the remark this answers is not in the document]
Comment (lzrd) [reply drv/lpc55-update-server/src/images.rs:32]: This should be using [ImageVectorsLpc55](https://github.com/oxidecomputer/hubris/blob/558ef2d6e6e961515bc3905947fe4f5b6390d08a/lib/lpc55-rot-startup/src/images.rs#L19) which should be pulled out and made more available.

**Review by labbott — APPROVED** [review]
